# STA 220: Midterm

### Worth 200 Points

Winter 2020

## Information

After the colons (in the same line) please write just your first name, last name, and the 9 digit student ID number below.

First Name: **Chenghan**

Last Name: **Sun**

Student ID: **915030521**

## Instructions

Please print your answer notebook to pdf (make sure that it is not unnecessarily long due to long output) and submit as the homework solution with your zip file.

For readability you,

- MUST add cells in between the exercise statements and add answers within them and
- MUST NOT modify the existing cells, particularly not the problem statement
- you may add multiple cells between exercise cells

To make markdown, please switch the cell type to markdown (from code) - you can hit 'm' when you are in command mode - and use the markdown language.  For a brief tutorial see: https://daringfireball.net/projects/markdown/syntax

## Trans-Atlantic Slave Trade

In this homework, we will uncover some of the numbers behind the Trans-atlantic slave trade (TAST), also known as the middle passage, that brought African slaves to the Americas.  The middle passage is reported to have forcibly migrated over 10 million Africans to the Americas over a roughly 3 century time span.  Many aspects of the TAST is little known by most people, such as the countries that constituted this network of slave ships, the regions from which the slaves were taken, and the number of slaves captured from Africa.

This last number is especially important since the number of slaves taken from Africa can impact other estimates that result from this.  For example, when estimating the population of Africa in a given decade, demographers will use population growth models and more recent census data.  For example, there are roughly X number of people in Africa and such populations tend to grow at rate M.  Then if we want to calculate the population one century ahead then we just apply a simple formula that assumes that the population grows at this rate.  But if the population is being drained by the slave trade, then this number will tend to be underestimated because the growth rate is overestimated.  To account for this models need to take into account this drain on the population.

Throughout this homework you will need to follow the principles of graphical excellence and the grammar of graphics.  **Use only Plotnine for your graphics**, do not use Pyplot, Seaborn, or Plotly since they do not follow closely the grammar of graphics.  Be sure to include titles and necessary contextual captions.

**Warning:** The Trans-Atlantic Slave Trade remains one of the most horrific abuses of human rights in history.  This homework deals with the numbers behind this forced migration, please be aware that this is a sensitive topic for possibly yourself and others.   A suitable amount of respect and seriousness is required when dealing with this data.

In [1]:
import pandas as pd
import numpy as np
import sys
from plotnine import *

**Exercise 1.** The data.

1. Read in the Trans-Atlantic Slave Trade database with Pandas.  Hint: if you use the unix tool `file` you can find that this CSV is encoded with iso-8859-1 character set.  Make sure that all missing values are encoded as NaN.
2. Open up the pdf file: TAST_codebook.pdf which is the data dictionary for this and other related datasets.  Many of the variables in the codebook are not in this dataset because it is describing an updated dataset. Create a list where you describe the meaning of the columns of your imported dataframe.  You can group similar columns together when describing their rough meaning, such as ownera,...,ownerp are owners of the slave ships.

Throughout we will disregard all time variables other than year since they are unreliable.

In [2]:
#1 Read in the Trans-Atlantic Slave Trade database with Pandas
data_path = './tastdb-2010.csv'
df_tast = pd.read_csv(data_path, encoding='iso-8859-1', na_values=' ', low_memory=False)
df_tast.head(5)

voyageid  evgreen         shipname  national  natinimp  placcons  yrcons  \
0         1      1.0  Pastora de Lima       4.0       6.0       NaN     NaN   
1         2      1.0          Tibério       NaN       6.0       NaN     NaN   
2         3      1.0     Paquete Real       NaN       6.0       NaN     NaN   
3         4      1.0      Bom Caminho       NaN       6.0       NaN     NaN   
4         5      1.0       Benigretta       NaN       6.0       NaN     NaN   

   placreg  yrreg   rig  ...  sourcei  sourcej  sourcek sourcel sourcem  \
0      NaN    NaN  40.0  ...      NaN      NaN      NaN     NaN     NaN   
1      NaN    NaN  40.0  ...      NaN      NaN      NaN     NaN     NaN   
2      NaN    NaN  40.0  ...      NaN      NaN      NaN     NaN     NaN   
3      NaN    NaN  40.0  ...      NaN      NaN      NaN     NaN     NaN   
4      NaN    NaN  25.0  ...      NaN      NaN      NaN     NaN     NaN   

  sourcen sourceo sourcep sourceq sourcer  
0     NaN     NaN     NaN     NaN     NaN  
1     NaN     NaN     NaN     NaN     NaN  
2     NaN     NaN     NaN     NaN     NaN  
3     NaN     NaN     NaN     NaN     NaN  
4     NaN     NaN     NaN     NaN     NaN  

[5 rows x 98 columns]

In [3]:
# check data dimensions
print(f'TAST text file: rows of the dataset = {df_tast.shape[0]}; number of variables = {df_tast.shape[1]}')
df_tast.columns # check all the variables

TAST text file: rows of the dataset = 34941; number of variables = 98


Index(['voyageid', 'evgreen', 'shipname', 'national', 'natinimp', 'placcons',
       'yrcons', 'placreg', 'yrreg', 'rig', 'tonnage', 'tonmod', 'guns',
       'ownera', 'ownerb', 'ownerc', 'ownerd', 'ownere', 'ownerf', 'ownerg',
       'ownerh', 'owneri', 'ownerj', 'ownerk', 'ownerl', 'ownerm', 'ownern',
       'ownero', 'ownerp', 'fate', 'fate2', 'fate3', 'fate4', 'resistance',
       'ptdepimp', 'plac1tra', 'plac2tra', 'plac3tra', 'mjbyptimp', 'npafttra',
       'sla1port', 'adpsale1', 'adpsale2', 'mjslptimp', 'portret', 'yearam',
       'Date_dep', 'Date_buy', 'Date_leftAfr', 'Date_land1', 'Date_depam',
       'Date_end', 'voy1imp', 'voy2imp', 'captaina', 'captainb', 'captainc',
       'crew1', 'crew3', 'crewdied', 'slintend', 'ncar13', 'ncar15', 'ncar17',
       'tslavesd', 'slaximp', 'slaarriv', 'slas32', 'slas36', 'slas39',
       'slamimp', 'menrat7', 'womrat7', 'boyrat7', 'girlrat7', 'malrat7',
       'chilrat7', 'jamcaspr', 'vymrtimp', 'vymrtrat', 'sourcea', 'sourceb',
       '

Noticed the existance of some imputed variables, I decided to drop them as they would have no benefits for later data analysis.

    For these imputed variables, some of them were titled with 'imp' as last three characters:  
        natinimp  
        ptdepimp  
        mjbyptimp  
        mjslptimp  
        voy1imp  
        voy2imp  
        slaximp  
        slamimp  
        vymrtimp
        
    Other imputed variables used special names:  
        tonmod  
        fate2  
        fate3  
        fate4  
        yearam  
        menrat7  
        womrat7  
        boyrat7  
        girlrat7  
        malrat7  
        chilrat7  
        jamcaspr  
        vymrtrat

In [4]:
imp_var_index = [i for i, var in enumerate(df_tast.columns) if 'imp' == var[-3:]]
rat7_var_index = [i for i, var in enumerate(df_tast.columns) if 'rat7' == var[-4:]]
other_var_index = [i for i, var in enumerate(df_tast.columns) if 'tonmod' == var[-6:] or 'ate2' == var[-4:] 
                   or 'ate3' == var[-4:] or 'ate4' == var[-4:] or 'yearam' == var[-6:] 
                   or 'yearam' == var[-6:] or 'yearam' == var[-6:] or 'jamcaspr' == var[-8:] 
                   or 'vymrtrat' == var[-8:]]
joint_imp_var_index = imp_var_index + rat7_var_index + other_var_index
# print(joint_imp_var_index)
# print(len(joint_imp_var_index))
df_tast_noimp = df_tast.drop(df_tast.columns[joint_imp_var_index], axis=1) # drop imputed variables
# print(len(df_tast_noimp.columns))
df_tast_noimp.columns

Index(['voyageid', 'evgreen', 'shipname', 'national', 'placcons', 'yrcons',
       'placreg', 'yrreg', 'rig', 'tonnage', 'guns', 'ownera', 'ownerb',
       'ownerc', 'ownerd', 'ownere', 'ownerf', 'ownerg', 'ownerh', 'owneri',
       'ownerj', 'ownerk', 'ownerl', 'ownerm', 'ownern', 'ownero', 'ownerp',
       'fate', 'resistance', 'plac1tra', 'plac2tra', 'plac3tra', 'npafttra',
       'sla1port', 'adpsale1', 'adpsale2', 'portret', 'Date_dep', 'Date_buy',
       'Date_leftAfr', 'Date_land1', 'Date_depam', 'Date_end', 'captaina',
       'captainb', 'captainc', 'crew1', 'crew3', 'crewdied', 'slintend',
       'ncar13', 'ncar15', 'ncar17', 'tslavesd', 'slaarriv', 'slas32',
       'slas36', 'slas39', 'sourcea', 'sourceb', 'sourcec', 'sourced',
       'sourcee', 'sourcef', 'sourceg', 'sourceh', 'sourcei', 'sourcej',
       'sourcek', 'sourcel', 'sourcem', 'sourcen', 'sourceo', 'sourcep',
       'sourceq', 'sourcer'],
      dtype='object')

**Part 2: List of variables**:
1. Technical variables:    
    voyageid: Voyage identification number  
    evgreen: Voyage in 1999 CD-ROM


2. Data variables  
    shipname: Name of vessel  
    national: Country i n which ship registered  
    placcons: Place where vessel constructed  
    yrcons: Year of vessel’s construction  
    placreg: Place where vessel registered  
    yrreg: Year of vessel’s registration  
    rig: Rig of vessel  
    tonnage: Tonnage of vessel  
    guns: Guns mounted  
    ownera, ownerb, ownerc, ownerd, ownere, ownerf, ownerg, ownerh,owneri, ownerj, ownerk, ownerl, ownerm, ownern, ownero, ownerp: first to sixteenth owner of venture  
    fate: Particular outcome of voyage  
    resistance: African resistance  
    plac1tra, plac2tra, plac3tra: First to third place of slave purchase  
    npafttra: Port of call before Atlantic crossing  
    sla1port, adpsale1, adpsale2: First to third place of slave landing  
    portret: Place at which voyage ended  
    Date_dep: Date that voyage began  
    Date_buy: Date that slave purchase began  
    Date_leftAfr: Date that vessel left last slaving port  
    Date_land1: Date that slaves landed at first place  
    Date_depam: Date ship left on return voyage  
    Date_end: Date when voyage completed  
    captaina, captainb, captainc: First to third captain’s name  
    crew1: Crew at voyage outset  
    crew3: Crew at first landing of slaves  
    crewdied: Crew died during complete voyage  
    slintend: Slaves intended from first port of purchase  
    ncar13, ncar15, ncar17: Slaves carried from first to third port of purchase  
    tslavesd: Total slaves on board at departure from last slaving port  
    slaarriv: Total slaves arrived at first port of disembarkation  
    slas32, slas36, slas39: Slaves disembarked at first to third place  
    sourcea, sourceb, sourcec, sourced, sourcee, sourcef, sourceg, sourceh, sourcei, sourcej, sourcek, sourcel, sourcem, sourcen, sourceo, sourcep, sourceq, sourcer: First to eighteenth source of information  
    

**Exercise 2.** Estimating the total number of captives.

1. We will try to estimate the number of people captured into slavery and forced through the middle passage.  What variable would you use to estimate the total number of captives taken from Africa?  Let me call this variable Var A in this problem statement.   How much of the data for Var A is missing?
2. Create an initial estimate of the total number of captives taken from Africa by assuming that Var A is Missing Completely at Random.  You can simply divide the total count for the non-missing entries by the proportion of non-missing entries.  
3. What other variables do you expect to be associated with Var A and why, select 2 top possibilities?  Visualize these associations using an appropriate plot.  Do you trust the answer to 2.2?  Why or why not?

**Part 1:** I choose to use variable `tslavesd` as estimation for the total number of captives taken from Africa, as it describes the "total slaves on board at departure from last slaving port".

In [5]:
var_A = 'tslavesd'
df_tast[var_A].isnull().sum()

26734

In [6]:
# Part 2
num_tslavesd = df_tast[var_A].sum()
portion = (df_tast.shape[0] - df_tast[var_A].isnull().sum()) / df_tast.shape[0]
num_tslavesd / portion

11569160.328337599

From the estimation above, the initial estimate of the total number of captives taken from Africa = 11569160.

In [13]:
# Part 3
# find the correlation between Var_A and all variables 
corr_var_a = df_tast_noimp[df_tast_noimp.columns[:]].corr()[var_A][:]
print(corr_var_a[0:])
df_corr_vara = pd.DataFrame(corr_var_a).rename(columns={var_A: 'corr_var_A'})
df_corr_vara

voyageid     -0.125522
evgreen            NaN
national     -0.186564
placcons     -0.016980
yrcons        0.196200
placreg      -0.325114
yrreg         0.146558
rig           0.113026
tonnage       0.540108
guns          0.526388
fate         -0.156838
resistance   -0.077623
plac1tra      0.170587
plac2tra      0.326746
plac3tra      0.297660
npafttra     -0.023067
sla1port      0.246330
adpsale1      0.140201
adpsale2      0.331953
portret      -0.184949
crew1         0.522192
crew3         0.347957
crewdied      0.199745
slintend      0.831760
ncar13        0.810653
ncar15        0.572813
ncar17        0.702715
tslavesd      1.000000
slaarriv      0.937504
slas32        0.783066
slas36        0.623793
slas39        0.629987
Name: tslavesd, dtype: float64


corr_var_A
voyageid     -0.125522
evgreen            NaN
national     -0.186564
placcons     -0.016980
yrcons        0.196200
placreg      -0.325114
yrreg         0.146558
rig           0.113026
tonnage       0.540108
guns          0.526388
fate         -0.156838
resistance   -0.077623
plac1tra      0.170587
plac2tra      0.326746
plac3tra      0.297660
npafttra     -0.023067
sla1port      0.246330
adpsale1      0.140201
adpsale2      0.331953
portret      -0.184949
crew1         0.522192
crew3         0.347957
crewdied      0.199745
slintend      0.831760
ncar13        0.810653
ncar15        0.572813
ncar17        0.702715
tslavesd      1.000000
slaarriv      0.937504
slas32        0.783066
slas36        0.623793
slas39        0.629987

In [8]:
df_corr_vara = pd.DataFrame(corr_var_a).rename(columns={var_A: 'corr_var_A'})
index = [i for i in range(len(df_corr_vara))]
df_corr_vara['index'] = index
df_corr_vara
(ggplot(df_corr_vara, aes('variables', 'corr_var_A', fill='corr_var_A'))
 + geom_tile(aes(width=.95, height=.95))
 + geom_text(aes(label='corr_var_A'), size=10)
)

/usr/local/lib/python3.7/site-packages/plotnine/layer.py:517: MatplotlibDeprecationWarning: isinstance(..., numbers.Number)
  return not cbook.iterable(value) and (cbook.is_numlike(value) or
/usr/local/lib/python3.7/site-packages/plotnine/layer.py:517: MatplotlibDeprecationWarning: isinstance(..., numbers.Number)
  return not cbook.iterable(value) and (cbook.is_numlike(value) or


PlotnineError: "Could not evaluate the 'x' mapping: 'variables' (original error: name 'variables' is not defined)"

From the above plot, 

**Exercise 3.** Hot-deck imputation replaces missing data with similar entries.  We will use a stochastic hot-deck imputation method, which replaces the number captured into slavery (tslavesd) with a random voyage from the same decade.  (If there is only one voyage in that decade then ignore that record.)  Construct a function that performs this random imputation method, and outputs a Series with this new imputed variable.  Calculate the total imputed tslavesd variable 100 times by performing this random imputation method 100 times.  Report the standard error, which is the standard deviation of the total imputed, along with the mean of the 100 trials.

In [27]:
case = df_tast_noimp['Date_dep'].dropna()
case2 = df_tast_noimp['Date_end'].dropna()
print(case.max())
print(case.min())
print(case2.max())
print(case2.min())

9/9/1847
1/1/1676
9/9/1804
1/1/1669


In [21]:
def hot_deck_imp(df, col_name, new_col_name):
    """ stochastic hot-deck imputation
    replaces the number captured into slavery (tslavesd) with a random voyage from the same decade
    """
    new_var = []
    selectd_df = df[col_name]
    pool = selectd_var.dropna()
    new_df = selectd_df.apply(lambda x: np.random.choice(pool) if np.isnan(x) else x)
    new_var.extend(new_df)
    df[new_col_name] = new_var
    return df

def hot_deck_stat(df):
    """ Report the standard errors
    """
    return None

**Exercise 4.** The flag that the ships flew.

1. We want to understand the trends of the nationality of the slave ships (the flag that they flew under is in the national variable).  Subselect the values of `national` that have more than 100 voyages with that value.  
2. Create a DataFrame that filters out the voyages where `national` does not have one of these values.  You should be retaining voyages with only these most common values.
3. Create a variable, `flag`, that is a string of easily readable names for these values by looking them up in the pdf codebook.
4. Using Plotnine, plot the counts of the voyages by flag as a function of voyage year.  Think about how best to display the count of a voyage by year and then how should you be including the flag variable.
5. In this plot, what are the geometric elements and aesthetic mappings?  What other components of the grammar of graphics are you using?
6. Do you observe any abrupt changes in the patterns of these counts for a given flag?  Investigate the cause for this change (using Google, etc.).

In [ ]:
# 1 
national_count = df_tast['national'].value_counts()
national_count[national_count > 100]

In [ ]:
# 2
'''
nation_select = 
sub_voyage = df[df['national'].isin(nation_select)]
sub_voyage = common_voyage.reset_index(drop=True)
sub_voyage.head()
'''
flag_df = df_tast[df_tast['national'].isin(national_count[national_count > 100].index)]
flag_df

In [ ]:
# demo plot 
flag = ['Great Britain', 'Portugal', 'France', 'U.S.A', 'Spain', 'Netherlands', 'Brazil', 'Denmark']
# !!! replacing national 
(ggplot(flag_df, aes(x='yearam', fill='factor(national)'))
 + geom_col(stat='count')
 + facet_wrap('~national')
 + scale_fill_discrete(name='Nation', labels=flag)
 + labs(x = 'Year', y = 'Voyages Total Count')
)

**Exercise 5.** Looking at some of these ships.

1. Search for the slave ship mentioned in the following wikipedia article: https://en.wikipedia.org/wiki/Brookes_(ship) Hint: Look at all records of ships with 'Brook' in the name and try to match the characteristics to those described.  How many voyages for this ship are in the data (try to exclude ships with the same name)?
2. Create a variable that is True if there was a resistance (like a slave revolt) on the ship.  Plot the density of ships as a function of year with and without revolts and compare these distributions.
3. The movie Amistad was based on a real slave ship and slave uprising.  Read about it here: https://en.wikipedia.org/wiki/La_Amistad  Try to find this ship by searching for it by name and also searching for ships in the same 10 year period as this event with a slave resistance.  If you think you found it describe it, otherwise describe the events of another voyage that you did find.  

**Exercise 6.** Other patterns.

1. The arrival and departure locations are quite detailed.  Look in the appendix of the codebook for the location codes.  Make a coarser version of both arrival and departure port variables (select just the last departure and first arrival) so that for example,  
```
30000 Caribbean 36100 Martinique 36101 Fort-Royale
```
is just encoded as '3' or Caribbean.
2. Plot the trend of voyages as a function of arrival location.  What trends do you see?
3. Do the same for departure location.
4. Plot the ratio of captives that are men as a function of year.  Include a smoother to describe the over all trend.  Also include in the plot another possible confounding variable.
5. Describe the geoms, aesthetic mappings, and other aspects of the plot.